modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/Handler.scala

```diff
@@ -12,28 +12,45 @@ object Handler {
     */
   def handler(q: EventQueue): Process[Engine, QState] = {
 
-    def handleUserEvent(ue: UserEvent): Engine[QState] = ue match {
-      case Start => log("Output: Started") *> switch(q)(Status.Running)
-      case Pause => log("Output: Paused") *> switch(q)(Status.Waiting)
-      case AddExecution(pend) => log("Output: Adding Pending Execution") // TODO: Implement handler
-      case Poll => log("Output: Polling current state")
-      case Exit => log("Bye") *> close(q)
+    def handleUserEvent(ue: UserEvent): Engine[Unit] = ue match {
+      case Start              =>
+        log("Output: Started") *>
+        switch(q)(Status.Running)
+      case Pause              =>
+        log("Output: Paused") *>
+        switch(q)(Status.Waiting)
+      case AddExecution(pend) =>
+        log("Output: Adding Pending Execution") // TODO: Implement handler
+      case Poll               =>
+        log("Output: Polling current state")
+      case Exit               =>
+        log("Bye") *> close(q)
     }
 
-    def handleSystemEvent(se: SystemEvent): Engine[QState] = se match {
-      case (Completed(i, r)) => log("Output: Action completed") *> complete(i, r)
-      case (Failed(i, e)) => log("Output: Action failed") *> fail(q)(i, e)
-      case Executed => log("Output: Execution completed, launching next execution") *> next(q)
+    def handleSystemEvent(se: SystemEvent): Engine[Unit] = se match {
+      case (Completed(i, r)) =>
+        log("Output: Action completed") *>
+        complete(i, r)
+      case (Failed(i, e))    =>
+        log("Output: Action failed") *>
+        fail(q)(i, e)
+      case Executed          =>
+        log("Output: Execution completed, launching next execution") *>
+        next(q)
       // TODO: Closing to facilitate testing, in reality it shouldn't close
-      case Finished => log("Output: Finished") *> close(q)
+      case Finished          =>
+        log("Output: Finished") *>
+        switch(q)(Status.Completed) *>
+        close(q)
     }
 
-    receive(q) >>= (
+    receive(q).flatMap(
       ev => Process.eval(
-        ev match {
-          case EventUser(ue) => handleUserEvent(ue)
-          case EventSystem(se) => handleSystemEvent(se)
-        }
+        (ev match {
```

Comment:
From what I understood from @jluhrs both types of events will come from different `Process`es but at some point they will have to be merged into one.

The key to achieve consistent concurrency in spite of the asynchronicity of its actions is that in the end the `Engine` only needs to handle a single stream of input events.